dashboard/app/helpers/users_helper.rb

```diff
@@ -136,11 +137,46 @@ def user_summary(user)
     end
 
     unless exclude_level_progress
-      uls = user.user_levels_by_level(script)
-      paired_user_level_ids = PairedUserLevel.pairs(uls.values.map(&:id))
-      script_levels = script.script_levels
       user_data[:completed] = user.completed?(script)
-      user_data[:levels] = {}
+      progress = get_level_progress([user], script)
+      user_data[:levels] = progress[user.id]
+    end
+
+    user_data
+  end
+
+  # Get the level progress for the specified script and user
+  # @param {User[]} users
+  # @param {Script} script
+  # @return A set of level progress objects, keyed first by userId and then by
+  #   levelId. The possible fields in each progress object are status, result,
+  #   submitted, readonly_answers, paired.
+  def get_level_progress(users, script)
+    level_progress = {}
+
+    # We need to get users_levels (and paired user_levels) for each student. Get
+    # them all in advance, so that we can have 3 total db queries, instead of
+    # having 3 per user
+    if users.length == 1
+      # If we have a single user, do this so that the result is cached on the user
+      # instance
+      all_user_levels = users.first.user_levels.where(script_id: script.id)
+      user_levels_by_user = {users.first.id => all_user_levels}
+    else
+      all_user_levels = UserLevel.where(user_id: users.map(&:id), script_id: script.id)
+      user_levels_by_user = all_user_levels.group_by(&:user_id)
+    end
+
+    all_paired_user_levels = PairedUserLevel.pairs(all_user_levels.map(&:id))
```

Comment:
Worth noting that this call is responsible for 2 of our 3 db hits. It might be possible to combine those two into one.

Reply:
We might even be able to combine all these queries into one with some `joins` (or `left_outer_joins`), but it would take some experimenting. 

How frequently do we expect this to be called? Is it worth further optimization now? Reducing from 3 per user to 3 total is already a big win.

Reply:
`get_script_progress` gets called every time we load the script overview, or use the dropdown in our header. This PR doesn't do anything to make those scenarios better, since they're only calling with one user. However, optimizing from 3 to 1 (or even just 3 to 2 by making PairedUserLevel better) might have a reasonable affect.